File: org.kevoree.modeling.microframework/src/test/java/org/kevoree/modeling/microframework/test/json/JSONSaveTest.java
package org.kevoree.modeling.microframework.test.json;

import org.junit.Assert;
import org.junit.Test;
import org.kevoree.modeling.api.Callback;
import org.kevoree.modeling.api.ThrowableCallback;
import org.kevoree.modeling.microframework.test.cloud.CloudUniverse;
import org.kevoree.modeling.microframework.test.cloud.CloudModel;
import org.kevoree.modeling.microframework.test.cloud.CloudView;
import org.kevoree.modeling.microframework.test.cloud.Node;

/**
 * Created by duke on 10/16/14.
 */
public class JSONSaveTest {

    @Test
    public void escapeJsonTest() {
        CloudModel universe = new CloudModel();
        universe.connect(null);
        CloudUniverse dimension0 = universe.newUniverse();
        CloudView time0 = dimension0.time(0l);
        Node root = time0.createNode();
        time0.setRoot(root,null);
        root.setName("root\nhello");
        final String[] result = new String[1];
        time0.json().save(root,new Callback<String>() {
            @Override
            public void on(String model) {
                result[0] = model;
            }
        });
        Assert.assertEquals(result[0], "[\n" +
                "{\"@class\":\"org.kevoree.modeling.microframework.test.cloud.Node\",\"@uuid\":1,\"@root\":true,\"name\":\"root\\nhello\"}\n" +
                "]\n");

        CloudUniverse dimension1 = universe.newUniverse();
        CloudView time10 = dimension1.time(0l);
        time10.json().load("[\n" +
                "\t{\n" +
                "\t\t\"@class\": \"org.kevoree.modeling.microframework.test.cloud.Node\",\n" +
                "\t\t\"@uuid\": \"1\",\n" +
                "\t\t\"@root\": \"true\",\n" +
                "\t\t\"name\": \"root\\nhello\"\n" +
                "\t}\n" +
                "]\n",new Callback<Throwable>() {
            @Override
            public void on(Throwable throwable) {
                if (throwable != null) {
                    throwable.printStackTrace();
                }
            }
        });
        time10.json().save(root,new Callback<String>() {
            @Override
            public void on(String model) {
                result[0] = model;
            }
        });
        Assert.assertEquals(result[0], "[\n" +
                "{\"@class\":\"org.kevoree.modeling.microframework.test.cloud.Node\",\"@uuid\":1,\"@root\":true,\"name\":\"root\\nhello\"}\n" +
                "]\n");

    }

    @Test
    public void jsonTest() {
        CloudModel universe = new CloudModel();
        universe.connect(null);
        CloudUniverse dimension0 = universe.newUniverse();

        CloudView time0 = dimension0.time(0l);
        Node root = time0.createNode();
        time0.setRoot(root,null);
        root.setName("root");
        Node n1 = time0.createNode();
        n1.setName("n1");
        Node n2 = time0.createNode();
        n2.setName("n2");
        root.addChildren(n1);
        root.addChildren(n2);


        final String[] result = new String[1];
        time0.json().save(root,new Callback<String>() {
            @Override
            public void on(String model) {
                result[0] = model;
            }
        });
        String payloadResult = "[\n" +
                "{\"@class\":\"org.kevoree.modeling.microframework.test.cloud.Node\",\"@uuid\":1,\"@root\":true,\"name\":\"root\",\"children\":[2,3]},\n" +
                "{\"@class\":\"org.kevoree.modeling.microframework.test.cloud.Node\",\"@uuid\":2,\"name\":\"n1\",\"op_children\":[1]},\n" +
                "{\"@class\":\"org.kevoree.modeling.microframework.test.cloud.Node\",\"@uuid\":3,\"name\":\"n2\",\"op_children\":[1]}\n" +
                "]\n";


        Assert.assertEquals(result[0], payloadResult);

    }

}

File: org.kevoree.modeling.microframework.typescript/src/test/ts/org/kevoree/modeling/microframework/test/json/JSONSaveTest.ts
///<reference path="../../../../../junit/Test.ts"/>
///<reference path="../../../api/Callback.ts"/>
///<reference path="../../../api/ThrowableCallback.ts"/>
///<reference path="../../../api/data/MemoryKDataBase.ts"/>
///<reference path="../cloud/CloudDimension.ts"/>
///<reference path="../cloud/CloudUniverse.ts"/>
///<reference path="../cloud/CloudView.ts"/>
///<reference path="../cloud/Node.ts"/>
// TODO Resolve static imports
///<reference path="../../../../../junit/Assert.ts"/>

class JSONSaveTest {

  public jsonTest(): void {
    var universe: CloudUniverse = new CloudUniverse(new MemoryKDataBase());
    universe.newDimension({on:function(dimension0: CloudDimension){
    var time0: CloudView = dimension0.time(0);
    var root: Node = time0.createNode();
    time0.setRoot(root);
    root.setName("root");
    var n1: Node = time0.createNode();
    n1.setName("n1");
    var n2: Node = time0.createNode();
    n2.setName("n2");
    root.addChildren(n1);
    root.addChildren(n2);
    var result: string[] = new Array();
    time0.createJSONSerializer().serialize(root, {on:function(model: string, err: Throwable){
    result[0] = model;
    if (err != null) {
      err.printStackTrace();
    }
}});
    var payloadResult: string = "[\n" + "{\n" + "\t\"@meta\" : \"org.kevoree.modeling.microframework.test.cloud.Node\",\n" + "\t\"@uuid\" : \"1\",\n" + "\t\"@root\" : \"true\",\n" + "\t\"name\" : \"root\",\n" + "\t\"children\" : [\"2\",\"3\"],\n" + "}\n" + ",{\n" + "\t\"@meta\" : \"org.kevoree.modeling.microframework.test.cloud.Node\",\n" + "\t\"@uuid\" : \"2\",\n" + "\t\"name\" : \"n1\",\n" + "}\n" + ",{\n" + "\t\"@meta\" : \"org.kevoree.modeling.microframework.test.cloud.Node\",\n" + "\t\"@uuid\" : \"3\",\n" + "\t\"name\" : \"n2\",\n" + "}\n" + "]\n";
    assertEquals(result[0], payloadResult);
    var pathN2: string[] = [null];
    n2.path({on:function(p: string){
    pathN2[0] = p;
}});
    assertEquals("/children[name=n2]", pathN2[0]);
    var pathN1: string[] = [null];
    n1.path({on:function(p: string){
    pathN1[0] = p;
}});
    assertEquals("/children[name=n1]", pathN1[0]);
    var pathR: string[] = [null];
    root.path({on:function(p: string){
    pathR[0] = p;
}});
    assertEquals("/", pathR[0]);
}});
  }

}

